internal/resource/charmhub/charmhub.go

```diff
@@ -62,15 +75,75 @@ type CharmHubClient struct {
 
 // GetResource returns data about the resource including an io.ReadCloser
 // to download the resource.  The caller is responsible for closing it.
-func (ch *CharmHubClient) GetResource(req ResourceRequest) (ResourceData, error) {
+func (ch *CharmHubClient) GetResource(ctx context.Context, req ResourceRequest) (ResourceData, error) {
 	ch.logger.Tracef("GetResource(%s)", pretty.Sprint(req))
-	var data ResourceData
 
+	res, resourceURL, err := ch.getResourceDetails(ctx, req)
+	if err != nil {
+		return ResourceData{}, errors.Capture(err)
+	}
+
+	ch.logger.Tracef("Read resource %q from %q", res.Name, resourceURL)
+
+	tmpFile, err := os.CreateTemp("", "resource-")
+	if err != nil {
+		return ResourceData{}, errors.Capture(err)
+	}
+	defer func() {
+		// Always close the file, we no longer require to have it open for
+		// this purpose. Another process/method can take over the file.
+		tmpFile.Close()
+
+		// If the download was successful, we don't need to remove the file.
+		// It is the responsibility of the caller to remove the file.
+		if err == nil {
+			return
+		}
+
+		// Remove the temporary file if the download failed. If we can't
+		// remove the file, log a warning, so the operator can clean it up
+		// manually.
+		removeErr := os.Remove(tmpFile.Name())
+		if removeErr != nil {
+			ch.logger.Warningf("failed to remove temporary file %q: %v", tmpFile.Name(), removeErr)
+		}
+	}()
+
+	downloadResult, err := ch.download(ctx, resourceURL, tmpFile)
+	if err != nil {
+		return ResourceData{}, errors.Capture(err)
+	}
+	if downloadResult.SHA384 != res.Fingerprint.String() {
+		return ResourceData{}, errors.Errorf(
+			"%w: %q, got %q", ErrUnexpectedFingerprint, res.Fingerprint.String(), downloadResult.SHA384,
+		)
+	}
+	if downloadResult.Size != res.Size {
+		return ResourceData{}, errors.Errorf(
+			"%w: %q, got %q", ErrUnexpectedSize, res.Size, downloadResult.Size,
+		)
+	}
```

Comment:
We have the same thing in charm, but we moved it else where. I wonder if you can do the same... See: https://github.com/juju/juju/blob/44a4ffb15f12dceb2b93bba6969477a71d82a346/internal/charm/charmdownloader/downloader.go#L66-L113

Reply:
Have done, this makes testing it slightly easier. 

Note that we still need the raw charmhub client in the `resource/charmhub` package because a call to `Refresh` is also made there. This means we now have a CharmHubClient and a Downloader which takes a CharmHubClient.

I started to change the charmhub package to use dependency injection rather than building its dependencies directly in the New function, but the changes got very big very fast. It seems like it'd be worth doing in the future but I think it would be too much for this PR.